### ISSUE TITLE
Train a reward model based on flant5-xxl

### ISSUE THREAD

**Issue by thaumstrial** [LAION-AI/Open-Assistant#743]
Use [flant5-xxl](https://huggingface.co/docs/transformers/main/en/model_doc/t5) encoder to train reward-model based on human-feedback dataset [conala](https://conala-corpus.github.io/).

- code will be placed in model/reward/flant5-xxl
- contrast this model with other reward models

Current reward models just trained on the [openai/summarize-from-feedback](https://github.com/openai/summarize-from-feedback).
This can help reward model to support more models and datasets.

**Comment by ontocord** [comment]
Also please collect the datasets and list here too please. Please mix the datasets when training. webgpt, summarize, conala, nli, stackexchange, etc. thank you!

**Comment by ontocord** [comment]
One experiment to do is to train a RM on the dataset, then label new unlabled data and filter and then use the new dataset as a new RM training dataset.

**Comment by ontocord** [comment]
cc @Graverman-dev if you are interested in helping with the conala RM.

**Comment by ontocord** [comment]
Also - you can try encoder of UL2.

**Comment by thaumstrial** [comment]
A general idea

1. build a reward-model (take the prompt and answer and output a reward value) based on flant5-xxl encoder with a fully connected feedforward neural network to convert the encoder outputs into a single value. 
2. train this model on the dataset generated by adjusting and mixing the following datasets (make the reward value output by the model fit the reward value of manual annotation)
- [WebGPT](https://huggingface.co/datasets/openai/webgpt_comparisons): each question and answer get the same score as score_0 or score_1
- [summarize-from-feedback](https://openaipublic.blob.core.windows.net/summarize-from-feedback/website/index.html#/): each paragraph and summarize get a score calculated by accuracy, coherence, coverage, and overall.
- [conala](https://conala-corpus.github.io/): each prompt and answer will get a positive score
- [NLI](https://paperswithcode.com/task/natural-language-inference): entailment get positive score, contradiction get negative score, neutral get smaller negative score.
3. test this model on the dataset mentioned above, and compare this model with other reward-models (like [reward model-based on Instructor](https://github.com/LAION-AI/Open-Assistant/issues/77))

**Comment by ontocord** [comment]
Yes! exactly!

**Comment by ontocord** [comment]
conala - mined has a score already. You can use a topN. You will need to do some filtering tho. For length. Some answers are just not good.

**Comment by theblackcat102** [comment]
@thaumstrial  you have any preliminary results showing this work? We have many requests for different models but unless its better than ELECTRA discriminator model (webgpt accuracy > 60% without overfitting ), we don't think its worth the afford to do it.

Current reward model was train on a mix of summarize from human feedback, GPTJ synthetic dataset (OA community dataset) and webgpt human feedback dataset.

**Comment by thaumstrial** [comment]
@theblackcat102 Got it. When I finish the experiment, I will put the results here to consider whether more effort is needed.

**Comment by ontocord** [comment]
@thaumstrial - following up. You gave me interesting results in DM. Maybe we can share here so others know about it?

**Comment by maw501** [comment]
Is this still open and needing work? I'd be very happy to help if so.

**Comment by thaumstrial** [comment]
@maw501 My experiment is over

| Model with MLP| WebGPT Accuracy |
| ----------- | --------- |
| T5 - flan - small | 53.2%|
| T5 - flan - base |  48.3%|
| T5 - flan - large | 54.6%|
| T5 - flan - XXL | 47.9%|

Two calculation flows are used:

1. Concatenate $Q$and $A$into A sentence to get $S$, add vectors representing question and answer before $Q$and $A$respectively, and then use the model to encode $S$to get $S'$. Finally, $S_n'$and $S_p'$are given to the MLP for classification.

2. The model to answer after coding $A_p '$and $A_n' $1 after encoding padding to the problem of $Q '$as dimensions, using MLP $A_p' $, $A_n '$, and $Q $mapped to get consistent dimension $A_p' $, $A_n '$,' $, and $Q Then, the normalized $A_p "$and $A_n" $dot products the normalized $Q "$respectively to get $r_p$and $r_n$. These two values represent the correlation between the answer and the question. If $r_p$is greater than $r_n$, it indicates that the model prediction is correct.

The model is trained on the above dataset, and the accuracy of the model is tested on the WebGPT validation set.



An accuracy of around 50% indicates that the model is not working, that the model is not getting better as the number of parameters increases, and that the whole model structure and training method maybe wrong.

**Comment by thaumstrial** [comment]
@maw501 Do you have a better idea?

**Comment by maw501** [comment]
Hi @thaumstrial! 👋 

I read [your write-up](https://thaumstrial.github.io/post/qian-chang-t5-encoder-mo-gai-ban-jiang-li-mo-xing/) on how you trained the reward model and that it didn't work. I don't think I follow the what procedure you used exactly - were you trying to replicate the InstructGPT paper? It wasn't clear to me from reading your "Process" section but that's possibly my misunderstanding. In particular, what's the idea with the contrastive part?

However, I was curious as to why the method didn't work - do you have a possible explanation for this? 

I would expect it's possible to train a reward model to > 50% accuracy so would be happy to help here - do you have code you could share?

**Comment by thaumstrial** [comment]
@maw501 Hi! 👏 I tried to replicate the reward-model base on the InstructGPT paper. I want the reward-model to decide which of two responses to the same question was better based on an assumption: the features of the correct answer given by the encoder should be similar to the features of the question, and the incorrect answer should be different. So I used another MLP to compute the similarity between the question and the answer encoded by the encoder.

As for why the model doesn't work, I think the dataset is the main problem. Besides the dataset is not big enough, there maybe also some unnoticeable human errors in the manually labeled data that prevent the machine from learning correctly.

I modify the code directly on https://github.com/LAION-AI/Open-Assistant/tree/main/model/reward/instructor
The code is a bit messy though, I can send it to you if you need it.

I'm glad you're here to help, but I've given up on RLHF's attempts at this. Because as I said above, small mistakes made by humans are learned by machines, and large amounts of high quality manually labeled data are also very expensive. So I wanted to try models with completely unsupervised learning.

**Comment by maw501** [comment]
Nice, thanks - will take a look!
